core/os/device/remotessh/device.go

```diff
@@ -0,0 +1,216 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remotessh
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/google/gapid/core/app/layout"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	"golang.org/x/crypto/ssh/knownhosts"
+
+	"github.com/google/gapid/core/os/device"
+	"github.com/google/gapid/core/os/device/bind"
+)
+
+// Device extends the bind.Device interface with capabilities specific to
+// remote SSH clients
+type Device interface {
+	bind.Device
+	// PushFile will transfer the local file at sourcePath to the remote
+	// machine at destPath
+	PushFile(ctx context.Context, sourcePath, destPath string) error
+	// Forward will forward the local port to the remote port
+	Forward(ctx context.Context, local, device uint16) error
+	// RemoveForward removes a port forward from local
+	RemoveForward(ctx context.Context, local uint16) error
+	// MakeTempDir makes a temporary directory, and returns the
+	// path, as well as a function to call to clean it up.
+	MakeTempDir(ctx context.Context) (string, func(ctx context.Context), error)
+	// WriteFile writes the given file into the given location on the remote device
+	WriteFile(ctx context.Context, contents io.Reader, mode string, destPath string) error
+	// ForwardPort forwards the remote port. It automatically selects an open
+	// local port.
+	ForwardPort(ctx context.Context, remoteport int) (int, error)
```

Comment:
Do we need both?

Reply:
Sorry, that should have been cleaned up, there is no Forward any more.

Reply:
Would you mind if I opened a bug for this?